### PR TITLE
7 develop episode 2

### PR DESCRIPTION
**Pull request by laurasootes**
This PR:

Adds episode 2 for the NLP course. It contains various preprocessing steps:

By hand
- reading in the data
- getting sentence tokens
- getting word tokens

Using spacy
- lowercasing
- stopword removal
- lemmatization
- part-of-speech tagging

### CODE REVIEW

File: episodes/episode02.md [hunk 1]
```diff
@@ -1,43 +1,288 @@
...
 
-::::::::::::::::::::::::::::::::::::::::::::::::
+:::::::::::::::::::::::::::::::::::::::::::::::: objectives
 
-::::::::::::::::::::::::::::::::::::: objectives
+After following this lesson, learners will be able to:
 
-- Explain how to use markdown with The Carpentries Workbench
-- Demonstrate how to include pieces of code, figures, and nested challenge blocks
+- Explain what preprocessing means.
```
Comment (n400peanuts) [inline episodes/episode02.md:16]: type something

Comment (n400peanuts) [reply episodes/episode02.md:16]: ciao

File: episodes/episode02.md [hunk 2]
```diff
@@ -1,43 +1,306 @@
...
+import spacy
+from wordcloud import WordCloud, STOPWORDS, ImageColorGenerator
+```
+
+We load the pipeline that we are going to use which is `en_code_web_md`. This a pipeline that is pretrained to do a number of NLP tasks for English texts. It is important to realize that in many cases you can use a pretrained model such as this one. You then do not have to do any training of your data. This is very nice, because training a model requires a whole lot of data, that would have to be analyzed by hand before you can start. It also requires a lot of specific understanding of NLP and a lot of time, and often it is simply not neccesary. These available models are trained on a lot of data, and have very good accuracy.
+
+<em>Maybe something on when you do not want to use pretrained models and train of finetune.</em>
+
+```python
+nlp = spacy.load('en_code_web_md')
```
Comment (n400peanuts) [inline episodes/episode02.md:198]: `en_code_web_md` is a model that does not come as a standard in the spacy installation so it might be useful to add instructions on how to download it first (added those in the suggestion above)

also, there is a typo

```suggestion
nlp = spacy.load('en_core_web_md')
```

File: episodes/episode02.md [hunk 3]
```diff
@@ -1,43 +1,306 @@
...
+
+```Python
+import spacy
+from wordcloud import WordCloud, STOPWORDS, ImageColorGenerator
+```
+
+We load the pipeline that we are going to use which is `en_code_web_md`. This a pipeline that is pretrained to do a number of NLP tasks for English texts. It is important to realize that in many cases you can use a pretrained model such as this one. You then do not have to do any training of your data. This is very nice, because training a model requires a whole lot of data, that would have to be analyzed by hand before you can start. It also requires a lot of specific understanding of NLP and a lot of time, and often it is simply not neccesary. These available models are trained on a lot of data, and have very good accuracy.
+
+<em>Maybe something on when you do not want to use pretrained models and train of finetune.</em>
+
```
Comment (n400peanuts) [inline episodes/episode02.md:196]: instructions to download `en_core_web_md`:


```suggestion
First, ensure that the model is correctly installed in your environment. If not, this code below will do that for you:

```python
!python -m spacy download en_core_web_md
```

```

File: episodes/episode02.md [hunk 4]
```diff
@@ -1,43 +1,306 @@
...
+There are multiple python packages that can be used to for NLP, such as `Spacy`, `NLTK`, `Gensim` and `PyTorch`. Here we will be using the `Spacy` package.
+
+Let's first load a few packages that we will be using:
+
+```Python
+import spacy
+from wordcloud import WordCloud, STOPWORDS, ImageColorGenerator
+```
+
+We load the pipeline that we are going to use which is `en_code_web_md`. This a pipeline that is pretrained to do a number of NLP tasks for English texts. It is important to realize that in many cases you can use a pretrained model such as this one. You then do not have to do any training of your data. This is very nice, because training a model requires a whole lot of data, that would have to be analyzed by hand before you can start. It also requires a lot of specific understanding of NLP and a lot of time, and often it is simply not neccesary. These available models are trained on a lot of data, and have very good accuracy.
```
Comment (n400peanuts) [inline episodes/episode02.md:193]: corrected typo in the name of the model `core` vs `code`
```suggestion
We load the pipeline that we are going to use which is `en_core_web_md`. This a pipeline that is pretrained to do a number of NLP tasks for English texts. It is important to realize that in many cases you can use a pretrained model such as this one. You then do not have to do any training of your data. This is very nice, because training a model requires a whole lot of data, that would have to be analyzed by hand before you can start. It also requires a lot of specific understanding of NLP and a lot of time, and often it is simply not neccesary. These available models are trained on a lot of data, and have very good accuracy.
```

File: episodes/episode02.md [hunk 5]
```diff
@@ -1,43 +1,306 @@
...
+```python
+# Get word tokens
+for token in doc[0:6]:
+    print(token.text)
+```
+
+## Stop word removal
+If we want to get an idea of what the text is about we can visualize the word tokens in a word cloud to see which words are most common.
+
+PUT THIS  IN A SEPARATE FUNCTION THAT IS LOADED INTO THE NOTEBOOK
```
Comment (n400peanuts) [inline episodes/episode02.md:233]: I suggest removing the following sentence as this is meant to be read for the instructors. Perhaps this is something we can add to the instructors folder? not sure how would this work though

```suggestion
```

File: episodes/episode02.md [hunk 6]
```diff
@@ -1,43 +1,306 @@
...
+```python
+from wordcloud import WordCloud
```
Comment (n400peanuts) [inline episodes/episode02.md:235]: I suggest removing this import as you already have imported wordcloud above in the notebook

```suggestion
```

File: episodes/episode02.md [hunk 7]
```diff
@@ -1,43 +1,306 @@
...
+      wc = WordCloud(stopwords=sw).generate(str(doc))
+      plt.imshow(wc, interpolation='bilinear')
+      plt.axis("off")
+      return plt.show()
+```
+
+```python
+plot_wordcloud(doc = doc)
+```
+
```
Comment (n400peanuts) [inline episodes/episode02.md:247]: FYI We can add images in the md, it would help to the reader to know if they got the same output as you. Maybe worth considering if you have time?

File: episodes/episode02.md [hunk 8]
```diff
@@ -1,43 +1,306 @@
...
+It recognizes determiners, nouns, adpositions, and more. But we can also see that it is not perfect and mistakes are made. That is something important to remember; any model, pretrained or if you train it yourself: there are always mistakes in it.
+
+
+:::::::::::::::::::: challenge
+
+We have gone through all the data preprocessing techniques in this episode. Now that you know how to apply them all, let's see how their applications influence the final output.
+
+Choose two or more preprocessing steps and apply them in two different orders. For example, try removing stop words before applying lemmatisation and the reverse. Plot the output in a word cloud.
 
+::::::::: solution
```
Comment (n400peanuts) [inline episodes/episode02.md:295]: The challenge above is a proposal for an exercise -- would that work for you? is there anything else we could have? if so, might be useful to add a solution as well in the tag below

**Review by n400peanuts — CHANGES_REQUESTED** [review]
Ciao Laura this looks wonderful. I have left some suggestions and corrected typos but in general I think this covers exactly what we were meant to teach in this lesson, so well done!!!

my main suggestion is to frame at the beginning data preprocessing as a step that is dependent by the nlp task. This might help also in adding the exercise at the end? the rest are very minor suggestions!

File: episodes/episode02.md [hunk 9]
```diff
@@ -1,43 +1,288 @@
...
-teaching: 10
-exercises: 2
+title: "Episode 2: Preprocessing"
+teaching: 10 
+exercises: 0
 ---
+:::::::::::::::::::::::::::::::::::::::::::::::: questions
 
-:::::::::::::::::::::::::::::::::::::: questions 
+- What different types of preprocessing steps are there?
```
[the remark this answers is not in the document]
Comment (laurasootes) [reply episodes/episode02.md:8]: What did you mean to add here? It seems that the suggested addition is missing

File: episodes/episode02.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (laurasootes) [reply episodes/episode02.md:17]: isn't the original phrasing correct?

File: episodes/episode02.md [hunk 10]
```diff
@@ -1,43 +1,288 @@
...
 
-::::::::::::::::::::::::::::::::::::: objectives
+After following this lesson, learners will be able to:
 
-- Explain how to use markdown with The Carpentries Workbench
-- Demonstrate how to include pieces of code, figures, and nested challenge blocks
+- Explain what preprocessing means.
+- Explain why we need preprocessing.
+- Do various preprocessing steps: lowercasing, handling new lines, tokenizing, stop word removal, parts-of-speech tagging, stemming/lemmatizing such that they: 
+have a list of words in a piece of text arranged with their parts of speech and lemmas or stems, and with very frequent words removed
```
[the remark this answers is not in the document]
Comment (laurasootes) [reply episodes/episode02.md:20]: I have now added: "be able to apply and use a spacy pretrained model"

File: episodes/episode02.md [hunk 11]
```diff
@@ -1,43 +1,288 @@
...
 
 ::::::::::::::::::::::::::::::::::::::::::::::::
 
-> ## Learning Objectives
+# Preprocessing
+
```
[the remark this answers is not in the document]
Comment (laurasootes) [reply episodes/episode02.md:26]: I added you suggestions with some slight modification

File: episodes/episode02.md [hunk 12]
```diff
@@ -1,43 +1,288 @@
...
+In order to start analyzing out text we will first do some preprocessing. Preprocessing means that we take a number of steps to put our data in a form that is analyzable; that there is no noise in the data and that we have some features of the text.
+
+Examples of preprocessing steps are:
+
+- tokenization: this means splitting up the text into individual tokens. You can for example create sentence tokens or words tokens.
+- lowercasing
+- stop words removal, where you remove common words such as `the` or `a` that you would note need in some further analysis steps.
+- lemmatization: here you would get the lemma of each word. You would get the form in which you find the word in the dictionary, such as the singular version of a plural of a noun, or the first person present tense of a verb instead of the past principle. You are making sure that you do not get different versions of the same word: you would convert `words` into `word` and `talking` into `talk`
```
[the remark this answers is not in the document]
Comment (laurasootes) [reply episodes/episode02.md:34]: I agree that it is confusion. I think that it might be better to omit stemming and only include lemmatization here?

File: episodes/episode02.md [hunk 13]
```diff
@@ -1,43 +1,288 @@
...
+- part of speech tagging. This means that you identify what type of word each is; such as nouns and verbs.
```
[the remark this answers is not in the document]
Comment (laurasootes) [reply episodes/episode02.md:35]: We can also just leave it for now and see if others give comments on this

File: episodes/episode02.md (hunk 2 above)
Comment (laurasootes) [reply episodes/episode02.md:198]: I also added some stuff on pretrained models from Hugging Face

**Review by n400peanuts — APPROVED** [review]
